File: screen-switcher-sample/src/main/java/com/jaynewstrom/screenswitchersample/MainActivity.java
package com.jaynewstrom.screenswitchersample;

import android.app.Activity;
import android.os.Bundle;
import android.view.MotionEvent;

import com.jaynewstrom.concrete.Concrete;
import com.jaynewstrom.concrete.ConcreteWall;
import com.jaynewstrom.screenswitcher.ScreenSwitcher;
import com.jaynewstrom.screenswitcher.ScreenSwitcherFactory;
import com.jaynewstrom.screenswitcher.ScreenSwitcherPopHandler;
import com.jaynewstrom.screenswitcher.ScreenSwitcherState;
import com.jnewstrom.screenswitcher.dialoghub.DialogHub;

import javax.annotation.Nullable;
import javax.inject.Inject;

public final class MainActivity extends Activity implements ScreenSwitcherPopHandler {

    @Inject ScreenSwitcherState screenSwitcherState;
    @Inject ScreenManager screenManager;
    @Inject DialogHub dialogHub;

    private ScreenSwitcher activityScreenSwitcher;
    private ConcreteWall<MainActivityComponent> activityWall;
    private PopCompleteHandler popCompleteHandler;

    @Override protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ConcreteWall<ApplicationComponent> foundation = Concrete.findWall(getApplicationContext());
        activityWall = foundation.stack(new MainActivityBlock(foundation.getComponent()));
        activityWall.getComponent().inject(this);
        activityScreenSwitcher = ScreenSwitcherFactory.activityScreenSwitcher(this, screenSwitcherState, this);
        screenManager.take(activityScreenSwitcher);
        dialogHub.attachActivity(this);
        dialogHub.restoreState();
    }

    @Override protected void onDestroy() {
        super.onDestroy();
        screenManager.drop(activityScreenSwitcher);
        dialogHub.dropActivity(this);
        if (popCompleteHandler != null) {
            popCompleteHandler.popComplete();
        }
        if (isFinishing()) {
            activityWall.destroy();
        } else {
            dialogHub.saveState();
        }
    }

    @Override public boolean dispatchTouchEvent(MotionEvent ev) {
        return activityScreenSwitcher.isTransitioning() || super.dispatchTouchEvent(ev);
    }

    @Override public void onBackPressed() {
        if (!activityScreenSwitcher.isTransitioning() && !isFinishing()) {
            screenManager.pop();
        }
    }

    @Override public Object getSystemService(String name) {
        if (Concrete.isService(name)) {
            return activityWall;
        }
        return super.getSystemService(name);
    }

    @Override public void onLastScreenPopped(PopCompleteHandler popCompleteHandler) {
        this.popCompleteHandler = popCompleteHandler;
    }
}

File: screen-switcher-sample/src/main/java/com/jaynewstrom/screenswitchersample/MainActivity.kt
package com.jaynewstrom.screenswitchersample

import android.app.Activity
import android.os.Bundle
import android.view.MotionEvent

import com.jaynewstrom.concrete.Concrete
import com.jaynewstrom.concrete.ConcreteWall
import com.jaynewstrom.screenswitcher.ScreenSwitcher
import com.jaynewstrom.screenswitcher.ScreenSwitcherFactory
import com.jaynewstrom.screenswitcher.ScreenSwitcherPopHandler
import com.jaynewstrom.screenswitcher.ScreenSwitcherState
import com.jnewstrom.screenswitcher.dialoghub.DialogHub
import javax.inject.Inject

class MainActivity : Activity(), ScreenSwitcherPopHandler {
    @Inject internal lateinit var screenSwitcherState: ScreenSwitcherState
    @Inject internal lateinit var screenManager: ScreenManager
    @Inject internal lateinit var dialogHub: DialogHub

    private lateinit var activityScreenSwitcher: ScreenSwitcher
    private lateinit var activityWall: ConcreteWall<MainActivityComponent>
    private var popCompleteHandler: ScreenSwitcherPopHandler.PopCompleteHandler? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val foundation = Concrete.findWall<ConcreteWall<ApplicationComponent>>(applicationContext)
        activityWall = foundation.stack(MainActivityBlock(foundation.component))
        activityWall.component.inject(this)
        activityScreenSwitcher = ScreenSwitcherFactory.activityScreenSwitcher(this, screenSwitcherState, this)
        screenManager.take(activityScreenSwitcher)
        dialogHub.attachActivity(this)
        dialogHub.restoreState()
    }

    override fun onDestroy() {
        super.onDestroy()
        screenManager.drop(activityScreenSwitcher)
        dialogHub.dropActivity(this)
        popCompleteHandler?.popComplete()
        if (isFinishing) {
            activityWall.destroy()
        } else {
            dialogHub.saveState()
        }
    }

    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
        return activityScreenSwitcher.isTransitioning || super.dispatchTouchEvent(ev)
    }

    override fun onBackPressed() {
        if (!activityScreenSwitcher.isTransitioning && !isFinishing) {
            screenManager.pop()
        }
    }

    override fun getSystemService(name: String): Any? {
        return if (Concrete.isService(name)) {
            activityWall
        } else {
            super.getSystemService(name)
        }
    }

    override fun onLastScreenPopped(popCompleteHandler: ScreenSwitcherPopHandler.PopCompleteHandler) {
        this.popCompleteHandler = popCompleteHandler
    }
}
